### ISSUE TITLE
Nextcloud 12.0.0 caches complete URLs in redis

### ISSUE THREAD

**Issue by golderweb** [nextcloud/server#5329]
### Steps to reproduce
1. Nextcloud Installation (manual update) from 11.0.3 with archive https://download.nextcloud.com/server/releases/nextcloud-12.0.0.tar.bz2 at cmdline
2. Navigating to Nextcloud page in browser, unfortunally without http**s**
3. Switching to **https**

### Expected behaviour
Switching from http to https has no effect on the page and everything works just as it should.

### Actual behaviour
Obviously Nextcloud 12.0.0 caches some full URLs in Redis-Storage. After switching from http to https the browser complains about Cross-Origin-Policy Violations because at least the background image URL stays http.
Much more confusing after login try, there occurs an Error while loading files page and NC is redirecting back to the login page.

The same problem also occurs when using the same NC Installation with different domains.

Stopping and restarting the Redis service fixes the problem (for the first used Protocol/Domain combination called after restart).

### Server configuration

**Operating system**:
CentOS 6
**Web server:**
Apache 2.2
**Database:**
MySQL 5.1.73
**PHP version:**
PHP 5.6.3

**Nextcloud version:** (see Nextcloud admin page)
12.0.0
**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated manualy from 11.0.3
**Where did you install Nextcloud from:**
https://download.nextcloud.com/server/releases/nextcloud-12.0.0.tar.bz2

**Signing status:**
No errors have been found.

**List of activated apps:**
Enabled:
  - activity: 2.5.2
  - bruteforcesettings: 1.0.2
  - calendar: 1.5.2
  - comments: 1.2.0
  - contacts: 1.5.3
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - federation: 1.2.0
  - files: 1.7.2
  - files_external: 1.3.0
  - files_sharing: 1.4.0
  - files_texteditor: 2.4.1
  - files_videoplayer: 1.1.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - password_policy: 1.2.2
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharebymail: 1.2.0
  - survey_client: 1.0.0
  - systemtags: 1.2.0
  - theming: 1.3.0
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - workflowengine: 1.2.0

Disabled:
  - admin_audit
  - encryption
  - files_pdfviewer
  - files_trashbin
  - files_versions
  - firstrunwizard
  - gallery
  - user_external
  - user_ldap

**Nextcloud configuration:**
```
{
    "system": {
        "instanceid": "occ9no54rveu",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "oc.***1.de",
            "oc.***2.de",
            "oc-staging.***1.de"
        ],
        "datadirectory": "\/home\/***\/oc-data",
        "overwrite.cli.url": "https:\/\/oc.***1.de\/",
        "dbtype": "mysql",
        "version": "12.0.0.29",
        "dbname": "***_nextcloud_12.0.0",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "forceSSLforSubdomains": true,
        "mail_smtpmode": "php",
        "mail_from_address": "svc-owncloud",
        "mail_domain": "***1.de",
        "loglevel": 2,
        "theme": "",
        "maintenance": false,
        "filelocking.enabled": "true",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.local": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "\/home\/***\/.redis\/sock",
            "port": 0,
            "timeout": 0
        }
    }
}
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
- Dropbox

**Are you using encryption:** yes/no
no
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

### Client configuration
**Browser:**
Mozilla Firefox 53.0.3
**Operating system:**
Ubuntu 16.04

**Comment by blizzz** [comment]
@LukasReschke any idea? or who else to ping?

**Comment by juliushaertl** [comment]
I guess it would make sense to change the globalPrefix of the cache to also include the complete baseUrl instead of the webroot only. 

https://github.com/nextcloud/server/blob/fb4576752440e90f6e54588b707184290d74f1c0/lib/private/Server.php#L444-L444

**Comment by blizzz** [comment]
Yeah, sounds like the way to do it. Do you mind opening a PR?

**Comment by juliushaertl** [comment]
A patch is at #5429 Please help testing ￼:wink:

**Comment by MorrisJobke** [comment]
fixed via #5429  for Nextcloud 13.
